Provide example for CXF SOAP (#144)
apache/camel-quarkus-examples: cxf-soap/README.adoc

= Camel Quarkus CXF SOAP example

> :cq-example-description: An example that shows how to use Camel master component.

Author: This one is fixed already, i've noticed it when generating the examples.json.

TIP: More info about generating Java classes from WSDL can be found at https://quarkiverse.github.io/quarkiverse-docs/quarkus-cxf/dev/user-guide/first-soap-client.html#wsdl2java.

== Java first

On the other hand if we don't have the WSDL file, we can create SOAP service directly from Java interface (see `org.acme.cxf.soap.pojo.service.ContactService`) which is then processed by CXF (will create WSDL file as a benefit - how to obtain it will be shown later) and from that point we can implement the service in Camel fashion (see `org.acme.cxf.soap.pojo.MyPojoRouteBuilder`) as with "WSDL first" approach (but here we implement the logic in bean service - see `org.acme.cxf.soap.pojo.service.impl.ContactServiceInMemoryImpl` and not directly in route).

The exposed contact web service will enable five operations - _addContact_, _getContact_, _getContacts_, _updateContact_, _removeContact_.

TIP: If you would like to only generate WSDL from Java, you can follow https://quarkiverse.github.io/quarkiverse-docs/quarkus-cxf/dev/user-guide/generate-wsdl-from-java.html.

> ---

Reviewer: What is the purpose of this `---`?

Author: I wanted to have the `Binding (advanced)` in lower size then other chapters (so i've used `===`) but that didn't add the horizontal line so i've added it manually. Do you think having it at the same level as eg `== Java first` is OK ?

Reviewer: Ah, `---` is horizontal line? I did not know. I thought it is some leftover. No problem with having it there.

=== Binding (Advanced)

> For illustrating how other `wsdl2java` options could be applied to `quarkus.cxf.codegen.wsdl2java.additional-params` we have added `src/main/resources/binding.xml` which enables us to use `LocalDate` (more common in Java world) instead of default XML Date representation `XMLGregorianCalendar`. But in terms of this quickstart, it is not important at all and we could work with `XMLGregorianCalendar` as well.

Reviewer: ```suggestion
For illustrating how other `wsdl2java` options could be applied via `quarkus.cxf.codegen.wsdl2java.additional-params`, we have added a custom binding defined in `src/main/resources/binding.xml`. 
It instructs CXF to use `LocalDate` (more common in Java world) instead of default XML Date representation `XMLGregorianCalendar`. 
```

== Start in the Development mode

[source,shell]
----
$ mvn clean compile quarkus:dev
----

The above command compiles the project, starts the application and lets the Quarkus tooling watch for changes in your
workspace. Any modifications in your project will automatically take effect in the running application.

> To start playing with the example, go to section `Local playground`.
> 

Reviewer: ```suggestion
```

TIP: Please refer to the Development mode section of
https://camel.apache.org/camel-quarkus/latest/first-steps.html#_development_mode[Camel Quarkus User guide] for more details.

> == Playground [[playground]]

Reviewer: ```suggestion
[[playground]]
== Playground
```

We can first try to add some contact with:
[source,shell]
----
curl -X POST -H "Content-Type: text/xml;charset=UTF-8" -d @src/main/resources/requests/contact/add.xml http://localhost:8080/cxf/services/contact
----
Then verify it was added with:
[source,shell]
----
$ curl -X POST -H "Content-Type: text/xml;charset=UTF-8" -d @src/main/resources/requests/contact/getAll.xml http://localhost:8080/cxf/services/contact
----
Which should return:

[source,xml]
----
<soap:Envelope
	xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
	<soap:Body>
		<ns2:getContactsResponse
			xmlns:ns2="http://camel.apache.org/test/ContactService">
			<return>
				<contacts>
					<name>Lukas</name>
					<address>
						<city>New York</city>
						<street>Sky 1234</street>
					</address>
					<type>PERSONAL</type>
				</contacts>
			</return>
		</ns2:getContactsResponse>
	</soap:Body>

> </soap:Envelope>

Reviewer: It will be very hard to read like this.
Could you please re-indent with 2 or 4 spaces?

----

We can also test our customer service:

[source,shell]
----
$ curl -X POST -H "Content-Type: text/xml;charset=UTF-8" -d @src/main/resources/requests/customer/getByName.xml http://localhost:8080/cxf/services/customer
----

You can observe that we have hardcoded `test` name at SOAPBody part in `src/main/resources/requests/customer/getByName.soap` as follows:
[source, xml]
----
<cus:getCustomersByName>
    <name>test</name>
</cus:getCustomersByName>
----

> We can try to alter it to non-valid request (validation is enabled with `schema-validation-enabled=true` at `org.acme.cxf.soap.wsdl.MyWsdlRouteBuilder`). So try to change `test` -> `t` and invoke it again, you should see following exception:

Reviewer: ```suggestion
We can try to alter it to non-valid request (the validation is enabled with `schema-validation-enabled=true` in `org.acme.cxf.soap.wsdl.MyWsdlRouteBuilder`). 
For example, you can change `test` to `t`.
Once you invoke the service again, you should see the following exception:
```

[source, xml]
----
<soap:Envelope
	xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">
	<soap:Body>
		<soap:Fault>
			<faultcode>soap:Client</faultcode>
			<faultstring>Unmarshalling Error: cvc-minLength-valid: Value 't' with length = '1' is not facet-valid with respect to minLength '2' for type '#AnonType_namegetCustomersByName'. </faultstring>
		</soap:Fault>
	</soap:Body>
</soap:Envelope>

> ----

Reviewer: Narrower indentation please
